iocage_lib/ioc_fstab.py

```diff
@@ -61,41 +63,155 @@ def __init__(self, uuid, action, source, destination, fstype, fsoptions,
         appropriate methods.
         """
+        actions = ['add', 'remove', 'edit', 'replace']
+
+        if self.action not in actions:
+            raise RuntimeError("Type of operation not specified!")
 
         if self.action == "add":
+            self.__validate_fstab__([self.mount])
             self.__fstab_add__()
-            self.__fstab_mount__()
+
+            try:
+                self.__fstab_mount__()
+            except RuntimeError:
+                iocage_lib.ioc_common.logit({
+                    'level': 'WARNING',
+                    'message': 'Mounting entry failed, check \'mount\''
+                },
+                    _callback=self.callback,
+                    silent=self.silent
+                )
         elif self.action == "remove":
             dest = self.__fstab_remove__()
-            self.__fstab_umount__(dest)
+
+            try:
+                self.__fstab_umount__(dest)
+            except RuntimeError:
+                iocage_lib.ioc_common.logit({
+                    'level': 'WARNING',
+                    'message': 'Unmounting entry failed, check \'mount\''
+                },
+                    _callback=self.callback,
+                    silent=self.silent
+                )
         elif self.action == "edit":
             self.__fstab_edit__()
         elif self.action == "replace":
+            self.__validate_fstab__([self.mount])
             self.__fstab_edit__(_string=True)
             self.__fstab_mount__()
-        else:
-            raise RuntimeError("Type of operation not specified!")
+
+    def __read_fstab__(self):
+        with open(f"{self.iocroot}/jails/{self.uuid}/fstab", "r") as f:
+            for line in f:
+                yield line.rstrip()
+
+    def __get_fstab_dests_(self):
+        dests = []
+
+        for line in self.fstab:
+            source, destination, fstype, options, \
+                dump, _pass = line.split()[0:6]
+            dests.append(destination)
+
+        return dests
+
+    def __validate_fstab__(self, fstab, mode='single'):
+        dests = []
+        verrors = []
+        jail_root = f'{self.iocroot}/jails/{self.uuid}/root'
+
+        for line in fstab:
+            source, destination, fstype, options, \
+                dump, _pass = line.split()[0:6]
+            source = pathlib.Path(source)
+            missing_root = False
+            dest = pathlib.Path(destination)
+
+            if mode != 'all' and (
+                self.action == 'add' or self.action == 'replace'
+            ):
+                if destination in self.dests:
+                    verrors.append(f'Destination: {self.dest} already exists!')
+                    break
+
+                if jail_root not in self.dest:
+                    verrors.append(
+                        f'Destination: {self.dest} must include '
+                        f'jail\'s mountpoint! ({jail_root})'
+                    )
+                    break
```

Comment:
It would be similar here I think as William pointed out ?

Reply:
Same thing as I said to William above, there is no point if the jail's mountpoint isn't included to even see if the destination or anything else is valid.